libs/@local/hql/span/src/storage.rs

```diff
@@ -0,0 +1,82 @@
+use crate::{tree::SpanNode, Span, SpanId};
+
+/// A collection of spans within a single source file.
+///
+/// This struct is used to store information about multiple spans within a single source file.
+pub struct SpanStorage<S> {
+    arena: Vec<S>,
+}
+
+impl<S> SpanStorage<S> {
+    #[must_use]
+    pub const fn new() -> Self {
+        Self { arena: Vec::new() }
+    }
+}
+
+impl<S> SpanStorage<S>
+where
+    S: Span,
+{
+    #[expect(
+        clippy::cast_possible_truncation,
+        reason = "The arena is not expected to be larger than u32::MAX"
+    )]
+    fn next_id(&self) -> SpanId {
+        SpanId::new(self.arena.len() as u32)
+    }
+
+    pub fn insert(&mut self, span: S) -> SpanId {
+        let id = self.next_id();
+        self.arena.push(span);
+
+        id
+    }
+
+    #[must_use]
+    pub fn get(&self, span: SpanId) -> Option<&S> {
+        let index = span.value() as usize;
```

Comment:
I got curious, it's because that case is specifically allowed:

https://github.com/rust-lang/rust-clippy/blob/a411267453b7744ac909c04302fe866fd1d183d1/clippy_lints/src/casts/cast_possible_truncation.rs#L114

does not lint it, if it's from `u32 -> usize` which seems to be common enough to allow for truncation, because you usually don't target `u16` platforms, but there isn't a `usize::from(u32)` (because `usize` might be `u16` on embedded targets therefore isn't always possible)